lightningd/jsonrpc.c

```diff
@@ -575,14 +578,140 @@ struct json_stream *json_stream_fail(struct command *cmd,
 	return r;
 }
 
+static struct command_result *command_exec(struct json_connection *jcon,
+                                           struct command *cmd,
+                                           const char *buffer,
+                                           const jsmntok_t *request,
+                                           const jsmntok_t *params)
+{
+	struct command_result *res;
+
+	res = cmd->json_cmd->dispatch(cmd, buffer, request, params);
+
+	assert(res == &param_failed
+	       || res == &complete
+	       || res == &pending
+	       || res == &unknown);
+
+	/* If they didn't complete it, they must call command_still_pending.
+	 * If they completed it, it's freed already. */
+	if (res == &pending)
+		assert(cmd->pending);
+
+	list_for_each(&jcon->commands, cmd, list)
+		assert(cmd->pending);
+
+	return res;
+}
+
+/* A plugin hook to take over (fail/alter) RPC commands */
+struct rpc_command_hook_payload {
+	struct command *cmd;
+	const char *buffer;
+	const jsmntok_t *request;
+};
+
+static void rpc_command_hook_serialize(struct rpc_command_hook_payload *p,
+                                       struct json_stream *s)
+{
+	const jsmntok_t *method, *params;
+
+	method = json_get_member(p->buffer, p->request, "method");
+	params = json_get_member(p->buffer, p->request, "params");
+
+	json_object_start(s, "rpc_command");
+	json_add_string(s, "method", tal_fmt(tmpctx, "%.*s",
+	                                     method->end - method->start,
+	                                     p->buffer + method->start));
+	json_add_string(s, "params", tal_fmt(tmpctx, "%.*s",
+	                                     params->end - params->start,
+	                                     p->buffer + params->start));
+	json_object_end(s);
```

Comment:
This is effectively filtering out anything that isn't a `method` or `params`. As we discussed in Berlin it might be advantageous to deviate slightly from the JSON-RPC spec, and allow additional top-level keys in order to authenticate the caller and authorize an RPC call. In order to do this we could add the entire incoming JSON object as the payload for `rpc_command` instead of selectively copying things over.

Reply:
This could easily be achieved by using `json_add_tok` from the previous commit.

Reply:
Hmm yes, I was starting to work on a basic RPC locker plugin "à la" bitcoin-core (`lockrpc`, `unlockrpc`, `setrpcpassphrase`) but maybe we then want to have the plugin registering for the `rpc_command` hook to parse an additional field like a cookie auth (once again as in core) ? We could call this plugin `macaroon` !

Reply:
Yes, copy entire thing please, rather than cherrypicking.  This allows users to get creative.  I may regret that, of course!